sonar-duplications/src/main/java/org/sonar/duplications/index/PackedMemoryCloneIndex.java

```diff
@@ -104,31 +109,102 @@ public PackedMemoryCloneIndex(int hashBytes, int initialCapacity) {
     List<Block> result = new ArrayList<>();
     int realIndex = resourceIdsIndex[index];
     while (index < size && FastStringComparator.INSTANCE.compare(resourceIds[realIndex], resourceId) == 0) {
-      // extract block (note that there is no need to extract resourceId)
-      int offset = realIndex * blockInts;
+      result.add(getBlock(realIndex, resourceId));
+
+      index++;
+      realIndex = resourceIdsIndex[index];
+    }
+    return result;
+  }
+
+  private Block createBlock(int index, String resourceId, @Nullable ByteArray byteHash) {
+    int offset = index * blockInts;
+
+    if (byteHash == null) {
       int[] hash = new int[hashInts];
       for (int j = 0; j < hashInts; j++) {
         hash[j] = blockData[offset++];
       }
-      int indexInFile = blockData[offset++];
-      int firstLineNumber = blockData[offset++];
-      int lastLineNumber = blockData[offset++];
-      int startUnit = blockData[offset++];
-      int endUnit = blockData[offset];
-
-      Block block = blockBuilder
-        .setResourceId(resourceId)
-        .setBlockHash(new ByteArray(hash))
-        .setIndexInFile(indexInFile)
-        .setLines(firstLineNumber, lastLineNumber)
-        .setUnit(startUnit, endUnit)
-        .build();
-      result.add(block);
+      byteHash = new ByteArray(hash);
+    } else {
+      offset += hashInts;
+    }
 
-      index++;
-      realIndex = resourceIdsIndex[index];
+    int indexInFile = blockData[offset++];
+    int firstLineNumber = blockData[offset++];
+    int lastLineNumber = blockData[offset++];
+    int startUnit = blockData[offset++];
+    int endUnit = blockData[offset];
+
+    return blockBuilder
+      .setResourceId(resourceId)
+      .setBlockHash(byteHash)
+      .setIndexInFile(indexInFile)
+      .setLines(firstLineNumber, lastLineNumber)
+      .setUnit(startUnit, endUnit)
+      .build();
+  }
+
+  private Block getBlock(int index, String resourceId) {
+    return createBlock(index, resourceId, null);
+  }
+
+  private class ResourceIterator implements Iterator<ResourceBlocks> {
+    private int index = 0;
+
+    @Override
+    public boolean hasNext() {
+      return index < size;
     }
-    return result;
+
+    @Override
+    public ResourceBlocks next() {
+      if (!hasNext()) {
+        throw new NoSuchElementException();
+      }
+
+      String resourceId = resourceIds[resourceIdsIndex[index]];
+      List<Block> blocks = new ArrayList<>();
+
+      // while we are at the same resource, keep going
+      do {
+        blocks.add(getBlock(resourceIdsIndex[index], resourceId));
+        index++;
+      } while (hasNext() && FastStringComparator.INSTANCE.compare(resourceIds[resourceIdsIndex[index]], resourceId) == 0);
+
+      return new ResourceBlocks(resourceId, blocks);
+    }
+
+    @Override
+    public void remove() {
+      throw new UnsupportedOperationException();
+    }
+  }
+
+  public class ResourceBlocks {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png) Make this a "static" inner class. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://dory.sonarsource.com/coding_rules#rule_key=squid%3AS2694)